File: api/tts.py
"""
Vercel Serverless Function (Python)
TTS con Kyutai Delayed Streams Modeling
Usa la biblioteca open-source Moshi para TTS de alta calidad
"""
from flask import Flask, request, jsonify
import base64
import os
import json
import tempfile
import torch

app = Flask(__name__)

# Configuración global
KYUTAI_CONFIG = {
    'model_repo': 'kyutai/tts-1b-en_es',  # Modelo multilingüe
    'sample_rate': 24000,
    'temperature': 0.7,
    'top_p': 0.9,
    'language': 'es',
}

# Cache del modelo (se carga una vez y se reutiliza)
_model_cache = None

def load_kyutai_model():
    """Carga el modelo Kyutai TTS (con cache)"""
    global _model_cache
    
    if _model_cache is not None:
        return _model_cache
    
    try:
        # Importar moshi
        from moshi import models
        
        # Cargar modelo desde Hugging Face
        print(f"📦 Cargando modelo Kyutai: {KYUTAI_CONFIG['model_repo']}")
        
        model = models.load_tts_model(
            hf_repo=KYUTAI_CONFIG['model_repo'],
            device='cuda' if torch.cuda.is_available() else 'cpu'
        )
        
        _model_cache = model
        print(f"✅ Modelo Kyutai cargado en {model.device}")
        return model
        
    except ImportError:
        print("❌ Error: biblioteca 'moshi' no encontrada")
        raise Exception("Moshi library not installed")
    except Exception as e:
        print(f"❌ Error cargando modelo Kyutai: {str(e)}")
        raise

def synthesize_with_kyutai(text, language='es'):
    """Sintetiza texto con Kyutai TTS"""
    try:
        # Cargar modelo
        model = load_kyutai_model()
        
        # Preparar input
        # Kyutai soporta prompts de voz y texto para mejor control
        synthesis_input = {
            'text': text,
            'language': language,
            'temperature': KYUTAI_CONFIG['temperature'],
            'top_p': KYUTAI_CONFIG['top_p'],
        }
        
        # Sintetizar
        print(f"🎙️ Sintetizando: {len(text)} caracteres")
        audio_output = model.synthesize(**synthesis_input)
        
        # Convertir a WAV en memoria
        import io
        import soundfile as sf
        
        audio_buffer = io.BytesIO()
        sf.write(
            audio_buffer, 
            audio_output, 
            KYUTAI_CONFIG['sample_rate'],
            format='WAV'
        )
        audio_buffer.seek(0)
        
        return audio_buffer.read()
        
    except Exception as e:
        print(f"❌ Error en síntesis Kyutai: {str(e)}")
        raise

def handler(request):
    """Handler principal para Vercel"""
    # CORS
    if request.method == 'OPTIONS':
        headers = {
            'Access-Control-Allow-Origin': '*',
            'Access-Control-Allow-Methods': 'POST, OPTIONS',
            'Access-Control-Allow-Headers': 'Content-Type',
        }
        return ('', 204, headers)
    
    headers = {
        'Access-Control-Allow-Origin': '*',
        'Content-Type': 'application/json'
    }
    
    if request.method != 'POST':
        return (json.dumps({'error': 'Method not allowed'}), 405, headers)
    
    try:
        # Obtener texto del request
        data = request.get_json()
        text = data.get('text', '')
        language = data.get('language', 'es')
        
        if not text:
            return (json.dumps({'error': 'Text is required'}), 400, headers)
        
        # Limitar a 3000 caracteres para eficiencia
        if len(text) > 3000:
            text = text[:3000]
        
        print(f"📝 Request TTS: {len(text)} chars, lang={language}")
        
        # Sintetizar con Kyutai
        audio_data = synthesize_with_kyutai(text, language)
        
        # Convertir a base64
        audio_base64 = base64.b64encode(audio_data).decode('utf-8')
        
        result = {
            'audioContent': audio_base64,
            'provider': 'Kyutai DSM TTS',
            'model': KYUTAI_CONFIG['model_repo'],
            'language': language,
            'characters': len(text),
            'sample_rate': KYUTAI_CONFIG['sample_rate'],
            'format': 'wav'
        }
        
        print(f"✅ TTS exitoso: {len(audio_data)} bytes")
        return (json.dumps(result), 200, headers)
        
    except Exception as e:
        print(f'❌ Error TTS: {str(e)}')
        import traceback
        traceback.print_exc()
        
        # Devolver fallback en caso de error
        result = {
            'error': str(e),
            'fallback': True,
            'provider': 'Kyutai DSM TTS (error)'
        }
        return (json.dumps(result), 500, headers)

File: api/tts.js
/**
 * Vercel Serverless Function - TTS Proxy
 * Proxy HTTPS para Coqui TTS Server en services VM
 *
 * Endpoint: services VM (10.204.0.5:5002)
 * Tecnología: Coqui TTS / gTTS
 * Fallback: Web Speech API del navegador
 *
 * Actualizado: 2025-12-01
 */

export default async function handler(req, res) {
  // CORS headers
  res.setHeader('Access-Control-Allow-Origin', '*');
  res.setHeader('Access-Control-Allow-Methods', 'POST, OPTIONS');
  res.setHeader('Access-Control-Allow-Headers', 'Content-Type');

  // Handle OPTIONS (preflight)
  if (req.method === 'OPTIONS') {
    return res.status(204).end();
  }

  // Only accept POST
  if (req.method !== 'POST') {
    return res.status(405).json({ error: 'Method not allowed' });
  }

  try {
    const { text, language = 'es', voice_id } = req.body;

    if (!text) {
      return res.status(400).json({ error: 'Text is required' });
    }

    // Limitar caracteres (TTS tiene límites de procesamiento)
    const truncatedText = text.length > 3000 ? text.substring(0, 3000) : text;

    // URL del servidor TTS en services VM
    // Puerto 5002 según especificaciones de red VPC
    const TTS_URL = process.env.TTS_URL || 'http://34.175.255.139:5002/speak';

    console.log(`📝 Proxy TTS: ${truncatedText.length} caracteres -> services VM:5002`);

    // Reenviar request a la VM services
    const response = await fetch(TTS_URL, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        text: truncatedText,
        language: language,
        voice_id: voice_id || 'sofia'
      }),
      signal: AbortSignal.timeout(30000) // 30 segundos timeout
    });

    if (!response.ok) {
      throw new Error(`TTS server responded with status ${response.status}`);
    }

    // Si es audio binario, reenviarlo directamente
    if (response.headers.get('content-type')?.includes('audio')) {
      const audioBuffer = await response.arrayBuffer();
      res.setHeader('Content-Type', 'audio/mpeg');
      return res.status(200).send(Buffer.from(audioBuffer));
    }

    // Si es JSON (respuesta con URL o datos)
    const data = await response.json();
    console.log('✅ TTS exitoso desde services VM');

    return res.status(200).json({
      ...data,
      provider: 'Coqui TTS',
      vm: 'services'
    });

  } catch (error) {
    console.error('❌ Error en proxy TTS:', error.message);

    // Devolver fallback para que el frontend use Web Speech API
    return res.status(200).json({
      error: error.message,
      fallback: true,
      provider: 'Web Speech API',
      message: 'TTS server no disponible, usando síntesis del navegador'
    });
  }
}

